### ISSUE TITLE
Suggestion: Synchronous resolvers

### ISSUE THREAD

**Issue by joezappie** [feathersjs/feathers#2601]
I'm not sure if something like this has already been discussed so sorry if there's already a way to do this. I've read through a lot of old issues and the dove docs but didn't see anything for this scenario.

**Problem**
I'm often finding myself wanting to have a resolver property that needs the value from another property that has not yet finished. An example (albeit not a great one) is setting the initials of a user based on the user that gets resolved in another property. Since resolvers are all ran async, the initials property function is going to be called before an the users service is called and sets the user value in the object.

```
resolve({
  properties: {
     user: async function(value, object, context) {
       return app.service('users').get(value);
     },
     initials: function(value, object, context) {
       // This will return undefined as object.user not set yet
       return object.user.name.match(/(\b\S)?/g).join('').toUpperCase();
     }
  }
});
```

**Current working solution**
The current solution to this would be to just set those properties in the user resolve function:

```
resolve({
  properties: {
     user: async function(value, object, context) {
       const user = await app.service('users').get(value);
       object.initials = value.name.match(/(\b\S)?/g).join('').toUpperCase();;
       return user;
     },
  }
});
```

This is a very simple example with a simple workaround, but I've got more complex resolvers where this feels hacky. The first time I ever tried to do something like this I coded it like my first example before realizing that wouldn't work. With the way resolvers feels like it was designed is each property should have its own function which helps keep it simple and clear.

Optionally, you can define two resolvers and add them both to the hooks array in the order you want them ran. This also feels error prone as I've created a bunch of common reusable resolvers. Having to remember to call multiple resolvers anytime I use it means I'm going to forget or put them in the wrong order. 

**Possible improvements**

1. Allow resolve() to take an array of objects. It could then run through each sequentially. 
2. In the root object, have a property for additional resolvers to run. 
3. Allow an object to be passed to a property, where another resolver can be passed in:
```
resolve({
  properties: {
     user: {
       value: async function(value, object, context) {
         return app.service('users').get(value);
       },
       resolver {
         properties: {
           initials: function(value, object, context) {
             return object.user.name.match(/(\b\S)?/g).join('').toUpperCase();
           }
       }
     },
  }
});
```

4. Provide a another function for running Synchronous resolvers (not tested but general idea):
```
// Export as sequentialResolve
export default (resolvers = {}) => async (context) => {
  for (const config of resolvers) {
    await resolve(config);
  }
};


// And run as a hook
sequentialResolve([
  {
    properties: {
       user: async function(value, object, context) {
         return app.service('users').get(value);
       },
    }
  },
  {
    properties: {
       initials: function(value, object, context) {
         return object.user.name.match(/(\b\S)?/g).join('').toUpperCase();
       }
    }
  };
]);
```

I'm not sure how this then works with resolveResult or resolveQuery though as it expects a single resolve function.

**TL:DR**
Option 4 feels the best to me. Having this built in and documented instead of having to figure it out would be nice.

**Comment by daffl** [comment]
I was thinking about some of this as well and to me the most straightforward way that's already possible would be to chain two resolvers. This should avoid any synchonization issues and keep the behaviour consistent: 

```js
const resolveUser = resolve({
  properties: {
     user: {
       value: async function(value, object, context) {
         return app.service('users').get(value);
       }
     },
  }
})

const resolveInitials = resolve({
  properties: {
    initials: function(value, object, context) {
      return object.user.name.match(/(\b\S)?/g).join('').toUpperCase();
    }
  }
})

[
  resolveResult(resolveUser),
  resolveResult(resolveInitials)
] 
```

Maybe we can change it to allow the hooks to chain them like

```js
  resolveResult(resolveUser, resolveInitials)
```

Which I think is very close to option 4 you are suggesting.

**Comment by DaddyWarbucks** [comment]
I also remember talking about this a while back. I second the idea of `resolveResult` taking multiple (or an array) of resolvers.

**Comment by joezappie** [comment]
`resultResult` taking an array would work for my use case so that seems like the way to go. All the helpers that can accept resolve() objects should also then be able to take an array such as resolveQuery?
